src/wp-includes/abilities-api/class-wp-ability.php

```diff
@@ -0,0 +1,588 @@
+	 * The detailed ability description.
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $description;
+
+	/**
+	 * The optional ability input schema.
+	 *
+	 * @since 6.9.0
+	 * @var array<string,mixed>
+	 */
+	protected $input_schema = array();
+
+	/**
+	 * The optional ability output schema.
+	 *
+	 * @since 6.9.0
+	 * @var array<string,mixed>
+	 */
+	protected $output_schema = array();
+
+	/**
+	 * The ability execute callback.
+	 *
+	 * @since 6.9.0
+	 * @var callable( mixed $input= ): (mixed|\WP_Error)
+	 */
+	protected $execute_callback;
+
+	/**
+	 * The optional ability permission callback.
+	 *
+	 * @since 6.9.0
+	 * @var callable( mixed $input= ): (bool|\WP_Error)
+	 */
+	protected $permission_callback;
+
+	/**
+	 * The optional ability metadata.
+	 *
+	 * @since 6.9.0
+	 * @var array<string,mixed>
+	 */
+	protected $meta;
+
+	/**
+	 * The ability category (required).
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $category;
+
+	/**
+	 * Constructor.
+	 *
+	 * Do not use this constructor directly. Instead, use the `wp_register_ability()` function.
+	 *
+	 * @access private
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see wp_register_ability()
+	 *
+	 * @param string              $name The name of the ability, with its namespace.
+	 * @param array<string,mixed> $args An associative array of arguments for the ability. This should include:
+	 *                                  `label`, `description`, `category`, `input_schema`, `output_schema`,
+	 *                                  `execute_callback`, `permission_callback` and `meta`
+	 */
+	public function __construct( string $name, array $args ) {
+		$this->name = $name;
+
+		$properties = $this->prepare_properties( $args );
+
+		foreach ( $properties as $property_name => $property_value ) {
+			if ( ! property_exists( $this, $property_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					sprintf(
+						/* translators: %s: Property name. */
+						esc_html__( 'Property "%1$s" is not a valid property for ability "%2$s". Please check the %3$s class for allowed properties.' ),
+						'<code>' . esc_html( $property_name ) . '</code>',
+						'<code>' . esc_html( $this->name ) . '</code>',
+						'<code>' . esc_html( self::class ) . '</code>'
+					),
+					'6.9.0'
+				);
+				continue;
+			}
+
+			$this->$property_name = $property_value;
+		}
+	}
+
+	/**
+	 * Prepares and validates the properties used to instantiate the ability.
+	 *
+	 * Errors are thrown as exceptions instead of \WP_Errors to allow for simpler handling and overloading. They are then
+	 * caught and converted to a WP_Error when by WP_Abilities_Registry::register().
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see WP_Abilities_Registry::register()
+	 *
+	 * @param array<string,mixed> $args An associative array of arguments used to instantiate the class.
+	 * @return array<string,mixed> The validated and prepared properties.
+	 * @throws \InvalidArgumentException if an argument is invalid.
+	 *
+	 * @phpstan-return array{
```

Comment:
https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-1-parameters-that-are-arrays